File: gadgets/python/accumulate_and_recon.py
import numpy as np
import kspaceandimage as ki
import libxml2
import ismrmrd
from gadgetron import Gadget

class AccumulateAndRecon(Gadget):
    myBuffer = []
    myParameters = {}
    myCounter = 1
    mySeries = 1

    def process_config(self, conf):
        doc = libxml2.parseDoc(str(conf))
        context = doc.xpathNewContext()
        context.xpathRegisterNs("ismrm", "http://www.ismrm.org/ISMRMRD")
        self.myParameters["matrix_x"] = int((context.xpathEval("/ismrm:ismrmrdHeader/ismrm:encoding/ismrm:encodedSpace/ismrm:matrixSize/ismrm:x")[0]).content)
        self.myParameters["matrix_y"] = int((context.xpathEval("/ismrm:ismrmrdHeader/ismrm:encoding/ismrm:encodedSpace/ismrm:matrixSize/ismrm:y")[0]).content)
        self.myParameters["matrix_z"] = int((context.xpathEval("/ismrm:ismrmrdHeader/ismrm:encoding/ismrm:encodedSpace/ismrm:matrixSize/ismrm:z")[0]).content)
        self.myParameters["channels"] = int((context.xpathEval("/ismrm:ismrmrdHeader/ismrm:acquisitionSystemInformation/ismrm:receiverChannels")[0]).content)
        try:
            self.myParameters["slices"] = int((context.xpathEval("/ismrm:ismrmrdHeader/ismrm:encoding/ismrm:encodingLimits/ismrm:slice/ismrm:maximum")[0]).content)+1
        except:
            self.myParameters["slices"] = 1
        self.myParameters["center_line"] = int((context.xpathEval("/ismrm:ismrmrdHeader/ismrm:encoding/ismrm:encodingLimits/ismrm:kspace_encoding_step_1/ismrm:center")[0]).content)

        self.myBuffer = (np.zeros((self.myParameters["channels"],self.myParameters["slices"],self.myParameters["matrix_z"],self.myParameters["matrix_y"],(self.myParameters["matrix_x"]>>1)))).astype('complex64')

    def process(self, acq, data):
        line_offset = (self.myParameters["matrix_y"]>>1)-self.myParameters["center_line"];
        self.myBuffer[:,acq.idx.slice,acq.idx.kspace_encode_step_2,acq.idx.kspace_encode_step_1+line_offset,:] = data

        if (acq.flags & (1<<7)): #Is this the last scan in slice
            image = ki.ktoi(self.myBuffer,(2,3,4))
            image = image * np.product(image.shape)*100 #Scaling for the scanner
            #Create a new image header and transfer value
            img_head = ismrmrd.ImageHeader()
            img_head.channels = acq.active_channels
            img_head.slice = acq.idx.slice
            img_head.matrix_size[0] = self.myBuffer.shape[4]
            img_head.matrix_size[1] = self.myBuffer.shape[3]
            img_head.matrix_size[2] = self.myBuffer.shape[2]
            img_head.position = acq.position
            img_head.read_dir = acq.read_dir
            img_head.phase_dir = acq.phase_dir
            img_head.slice_dir = acq.slice_dir
            img_head.patient_table_position = acq.patient_table_position
            img_head.acquisition_time_stamp = acq.acquisition_time_stamp
            img_head.image_index = self.myCounter;
            img_head.image_series_index = self.mySeries;

            self.myCounter += 1
            if self.myCounter > 5:
                    self.mySeries += 1
                    self.myCounter = 1

            #Return image to Gadgetron
            self.put_next(img_head,image)
            
        #print "Returning to Gadgetron"
        return 0 #Everything OK

File: gadgets/matlab/accumulate_and_recon.m
classdef accumulate_and_recon < handle & BaseGadget

    properties

        image_num;
        series_num;
        center_line;
        accumulation;
        
    end

    methods

        function g = config(g)
            fprintf('The resonance frequency is %d\n', g.xml.getExperimentalConditions().getH1ResonanceFrequencyHz());
            nx = g.xml.getEncoding().get(0).getEncodedSpace().getMatrixSize().getX();
            ny = g.xml.getEncoding().get(0).getEncodedSpace().getMatrixSize().getY();
            % for 2D sequences the number of getZ breaks
            try
              nz = g.xml.getEncoding().get(0).getEncodedSpace().getMatrixSize().getZ();
            catch
	      nz =1;
            end
            % the number of receiver channels is optional
            try
                % this is the only cast from java.lang.Integer that works in Matlab
                nc = double(g.xml.getAcquisitionSystemInformation().getReceiverChannels());
            catch
	        nc = 1;
            end
            % the number of slices is optional
            try
                ns = g.xml.getEncoding().get(0).getEncodingLimits().getSlice().getMaximum() + 1;
            catch
	        ns = 1;
            end

            g.center_line = g.xml.getEncoding().get(0).getEncodingLimits().getKspaceEncodingStep1().getCenter();
            g.accumulation = zeros(nx, ny, nz, ns, nc);
            g.image_num = 0;   % todo this needs to be static or global...
            g.series_num = 0;  % todo this needs to be static or global...
        end

        function g = process(g, head, data)
            % stuff the line
            line_offset = floor(size(g.accumulation,2)/2) - g.center_line;
            kyind = head.idx.kspace_encode_step_1 + line_offset + 1;
            kzind = head.idx.kspace_encode_step_2 + 1;
            slind = head.idx.slice + 1;
            %fprintf('  offset = %d, center = %d, index = %d\n', line_offset, g.center_line, kyind);

            g.accumulation(:, kyind, kzind, slind, :) = data;

            % At the end of the acquisition, reconstruct the slice
            if (head.flagIsSet(head.FLAGS.ACQ_LAST_IN_SLICE))
                img_head = ismrmrd.ImageHeader;
                img_head.channels = head.active_channels;
                img_head.slice = head.idx.slice;
                % set the matrix size
                % set one element at a time to not break the type (uint16) of matrix_size
     	        img_head.matrix_size(1) = size(g.accumulation,1); % nx
     	        img_head.matrix_size(2) = size(g.accumulation,2); % ny
     	        img_head.matrix_size(3) = size(g.accumulation,3); % nz

                img_head.position = head.position;
                img_head.read_dir = head.read_dir;
                img_head.phase_dir = head.phase_dir;
                img_head.slice_dir = head.slice_dir;
                img_head.patient_table_position = head.patient_table_position;
                img_head.acquisition_time_stamp = head.acquisition_time_stamp;
                img_head.image_index = g.image_num;
                img_head.image_series_index = g.series_num;

		img_data = squeeze(g.accumulation(:,:,:,slind,:));
                img_data = fftshift(ifftn(fftshift(img_data)));
                imagesc(abs(img_data(:,:,1,1))); axis image; axis square;
		pause(2)
                close()

                g.putQ(img_head, img_data);
                %fprintf('Put on Queue %d, type = %d\n',length(g.Q),g.Q{1}.type);

            end

        end

    end
end
